pkg/aspect/inventory.go

```diff
@@ -14,36 +14,74 @@
 
 package aspect
 
-import (
-	"istio.io/mixer/pkg/config"
+// APIMethod constants are used to refer to the methods handled by api.Handler
+type APIMethod int
+
+// Supported API methods
+const (
+	CheckMethod APIMethod = iota
+	ReportMethod
+	QuotaMethod
 )
 
-// APIBinding associates an aspect with an API method
-type APIBinding struct {
-	Aspect Manager
-	Method config.APIMethod
-}
+// Kind of aspect
+type Kind int
+
+// Supported kinds of aspects
+const (
+	AccessLogsKind Kind = iota
+	ApplicationLogsKind
+	DenialsKind
+	ListsKind
+	MetricsKind
+	QuotasKind
+)
 
 // Name of all supported aspect kinds.
 const (
-	AccessLogKind = "access-logs"
-	DenyKind      = "denials"
-	ListKind      = "lists"
-	LogKind       = "application-logs"
-	MetricKind    = "metrics"
-	QuotaKind     = "quotas"
+	AccessLogsKindName      = "access-logs"
+	ApplicationLogsKindName = "application-logs"
+	DenialsKindName         = "denials"
+	ListsKindName           = "lists"
+	MetricsKindName         = "metrics"
+	QuotasKindName          = "quotas"
 )
 
-// Inventory returns a manager inventory that contains
-// all available aspect managers
-func Inventory() []APIBinding {
-	// Update the following list to add a new Aspect manager
-	return []APIBinding{
-		{NewDenialsManager(), config.CheckMethod},
-		{NewListsManager(), config.CheckMethod},
-		{NewApplicationLogsManager(), config.ReportMethod},
-		{NewAccessLogsManager(), config.ReportMethod},
-		{NewQuotasManager(), config.CheckMethod},
-		{NewQuotasManager(), config.QuotaMethod},
-	}
+// KindNames maps from kinds to their names.
+var KindNames = map[Kind]string{
+	AccessLogsKind:      AccessLogsKindName,
+	ApplicationLogsKind: ApplicationLogsKindName,
+	DenialsKind:         DenialsKindName,
+	ListsKind:           ListsKindName,
+	MetricsKind:         MetricsKindName,
+	QuotasKind:          QuotasKindName,
+}
+
+// NamedKinds maps from kind names to kind enum.
+var NamedKinds = map[string]Kind{
+	AccessLogsKindName:      AccessLogsKind,
+	ApplicationLogsKindName: ApplicationLogsKind,
+	DenialsKindName:         DenialsKind,
+	ListsKindName:           ListsKind,
+	MetricsKindName:         MetricsKind,
+	QuotasKindName:          QuotasKind,
+}
+
+// ManagerInventory holds a set of aspect managers.
+type ManagerInventory map[APIMethod][]Manager
+
+// Inventory is the authoritative set of aspect managers used by the mixer.
+var Inventory = ManagerInventory{
```

Comment:
Why not use global constants? Makes for nicer code at the call site.

Using a function instead makes it so the caller can't change the actual variable, that's true. But since the function returns a reference to some underlying storage, the caller is still free to mutate the state. So unless we'd be copying the data before returning the function, there is no real difference between a function and a global constant.